### PR TITLE
BUG: boxcox data constancy check #5112 #10978

### PR DESCRIPTION
**Pull request by LaurynasMiksys**
Perform data constancy check only when `lmbda` parameter is not passed. There is no reason to fail execution when pre-fitted `lambda` value is passed to the transform. 

When running function at inference time (say, on a single value array), it should be able to perform desired transformation.

#### Reference issue
#5112 #10978

#### What does this implement/fix?
Performs additional check, if lmbda parameter is passed (not None), do not perform data constancy check.

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
@LaurynasMiksys, thanks for contributing this fix!  I agree, most of the checks that `boxcox` currently makes are only needed when lambda is not given.

Your change is fine, but in fact, we can be even more accepting of `x` values when lambda is given.  Instead of moving the check for constant input down, we can move this check

```
    if lmbda is not None:  # single transformation
        return special.boxcox(x, lmbda)
```
to the very beginning of the function.  That is, let `scipy.special.boxcox` handle *all* cases when `lambda` is given.  This allows the function to accept a scalar, an n-dimensional array, etc.

**Comment by LaurynasMiksys** [comment]
@WarrenWeckesser, thanks for reviewing! I do agree with your suggestion, except with one point - it should still check for positivity of the values. Otherwise, `scipy.special.boxcox` will crash when evaluating `log` of non-positive value.

**Comment by WarrenWeckesser** [comment]
Actually, instead of raising an exception, `scipy.special.boxcox` will return `nan` for invalid inputs, and `-inf` for the special case where the function is equivalent to `log(0)`:

```
In [209]: from scipy.special import boxcox

In [210]: boxcox([-1, 0, 1, 2], 2.5)
Out[210]: array([       nan, -0.4      ,  0.       ,  1.8627417])

In [211]: boxcox([-1, 0, 1, 2], 0)
Out[211]: array([       nan,       -inf, 0.        , 0.69314718])
```

If we make clear in the docstring that this is the behavior when lambda is given, I think it is OK for `scipy.stats.boxcox` to do the same.  It is only when the optimization process is being invoked--by not giving lambda--that we are much more noisy with the input validation.

Do any other devs have a preference?

**Comment by LaurynasMiksys** [comment]
On the second thought, I'm not entirely sure what would be the meaning of multivariate Box-Cox transformation.

One option would be to simply use scalar `lmbda` to transform n-dimensional array, disregarding the dimensions. But this would not transform multivariate distribution into multivariate normal. It would treat the values as if they are from 1-dimensional distribution.

Possibly there exists, but I'm not aware of, multivariate Box-Cox version. I'm not sure current implementation would handle it anyways. I would think, in that case `lmbda` would need to be an array itself.

In the former case, I would say it's better to notify the user that Box-Cox is performed on the array which is not 1-dimensional, rather than execute silently. User could easy transform the data into right shape on his/her end, if needed.

**Comment by WarrenWeckesser** [comment]
`scipy.special.boxcox` is not a multivariate Box-Cox transformation.  It is an element-wise operation, with numpy broadcasting of the arguments.  For example,
```
In [216]: boxcox([[1, 2], [3, 4]], [1, 5])
Out[216]: 
array([[  0. ,   6.2],
       [  2. , 204.6]])
```
is the same as
```
In [217]: np.array([[boxcox(1, 1), boxcox(2, 5)], [boxcox(3, 1), boxcox(4, 5)]])                     
Out[217]: 
array([[  0. ,   6.2],
       [  2. , 204.6]])
```

**Comment by LaurynasMiksys** [comment]
I see, thanks for clarification. 

Also, agree with your point that docstring should be made clearer with regards to this behaviour.

**Review by mdhaber — APPROVED** [review]
There was a change to `_normplot` that required the data to be positive. This would not have been correct for `yeojohnson_normplot`, so I added the condition that `method=='boxcox'`.

Otherwise, this makes the documentation more explicit and makes `boxcox` more flexible when `lmda` is provided, so LGTM.

File: scipy/stats/_morestats.py [hunk 1]
```diff
@@ -956,15 +956,27 @@ def boxcox(x, lmbda=None, alpha=None, optimizer=None):
     Parameters
     ----------
     x : ndarray
-        Input array.  Must be positive 1-dimensional.  Must not be constant.
-    lmbda : {None, scalar}, optional
+        Input array to be transformed.
+
+        If `lmbda` is not None, this is an alias of `special.boxcox`.
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:961]:
```suggestion
        If `lmbda` is not None, this is an alias of
        `scipy.special.boxcox`.
```

**Comment by mdhaber** [comment]
Thanks @LaurynasMiksys!

**Pull request merged**